File: LeetCode/1870/cpp/lc1870.cpp
#include <vector>
#include <cassert>
#include <cmath>

class Solution {
public: 
  int minSpeedOnTime(std::vector<int>& dist, double hour) {

    if (hour <= static_cast<double>(dist.size() - 1)) {
      // as each train takes at least 1 full hour to complete + embark on next one,
      // it is impossible to arrive on time if hour <= #trains - 1.
      return -1;
    }
    int low = 1; // lowest bound for optimal train speed.
    int high = 1e7; // highest bound
    while (low < high) {
      int mid = low + (high - low) / 2;
      if (this->canMakeOnTime(dist, hour, mid)) {
        // Midpoint can make it on time, making it an upper bound for true optimal speed.
        // (trains of midpoint speed reach target in desired time)
        high = mid;
      }
      else {
        // Midpoint can not make it on time, making it a lower bound for true optimal speed.
        // (trains of speed up to midpoint won't reach target in time)
        low = mid + 1;
      }
    }
    return low;
  }

private:
  bool canMakeOnTime(std::vector<int>& dist, double hour, double train_speed) {
    double travel_time = 0.0;
    for (auto it = dist.begin(); it != dist.end() - 1; it++) {
      travel_time += ceil(*it / train_speed);
    }
    travel_time += dist.back() / train_speed;
    bool ans = travel_time <= hour;
    return ans;
  }
};

int main() {
  Solution s{};
  std::vector<int> test1 = {1, 3, 2};
  assert(s.minSpeedOnTime(test1, 6.0) == 1);
  assert(s.minSpeedOnTime(test1, 2.7) == 3);
  assert(s.minSpeedOnTime(test1, 1.9) == -1);

  std::vector<int> input4 = {1, 1, 100'000};
  assert(s.minSpeedOnTime(input4, 2.01) == 100 * 100'000);

  std::vector<int> input5 = {1, 1};
  assert(s.minSpeedOnTime(input5, 1.0) == -1);

  std::vector<int> tc30 = {5, 3, 4, 6, 2, 2, 7};
  assert(s.minSpeedOnTime(tc30, 10.92) == 4);
} 

File: LeetCode/1870/golang/lc1870.go
package lc1870

import "math"

// Runtime complexity: O(n*log(1e7)) == O(n)
// Auxiliary space: O(1)
// Subjective level: medium
// Solved on: 2025-12-23
func minSpeedOnTime(dist []int, hour float64) int {
	n := len(dist)

	// as each train takes at least 1 full hour to complete its ride,
	// it is impossible to arrive on time if hour <= num_trains - 1:
	if hour <= float64(n-1) {
		return -1
	}

	canMakeOnTime := func(maxTime float64, trainSpeed float64) bool {
		travelTime := 0.0
		for d := 0; d < n-1; d++ {
			travelTime += math.Ceil(float64(dist[d]) / trainSpeed)
		}
		travelTime += float64(dist[n-1]) / trainSpeed
		return travelTime <= maxTime
	}

	low, high := 1, 10_000_000 // low and high boundaries for optimal train speed.
	for low < high {
		var mid int = low + (high-low)/2
		if canMakeOnTime(hour, float64(mid)) {
			// Midpoint speed makes the train reach target in time,
			// making it an upper bound for true optimal speed.
			high = mid
		} else {
			// Midpoint speed is insufficient to reach target in time,
			// making (midpoint+1) a lower bound for true optimal speed.
			low = mid + 1
		}
	}
	return low
}
